akka-stream/src/main/scala/akka/stream/impl/fusing/MapAsyncPartitioned.scala

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (C) 2022 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl.fusing
+
+import akka.stream._
+import akka.stream.ActorAttributes.SupervisionStrategy
+import akka.stream.Attributes.{ SourceLocation }
+import akka.stream.impl.{ Buffer => BufferImpl, PartitionedBuffer }
+import akka.stream.impl.Stages.DefaultAttributes
+import akka.stream.stage._
+import akka.util.OptionVal
+
+import scala.annotation.tailrec
+import scala.concurrent.Future
+import scala.util.{ Failure, Success, Try }
+import scala.util.control.NonFatal
+
+private[akka] object MapAsyncPartitioned {
+  def apply[In, Out, Partition](
+      parallelism: Int,
+      perPartition: Int,
+      partitioner: In => Partition,
+      f: (In, Partition) => Future[Out]): GraphStage[FlowShape[In, Out]] =
+    if (parallelism < 1) throw new IllegalArgumentException("parallelism must be at least 1")
+    else if (perPartition < 1) throw new IllegalArgumentException("perPartition must be at least 1")
+    else if (perPartition >= parallelism) {
+      // no point controlling per-partition, so just use a MapAsync, which has somewhat less overhead
+      val fin = { (x: In) =>
+        f(x, partitioner(x))
+      }
+      MapAsync(parallelism, fin)
+    } else new MapAsyncPartitioned(parallelism, perPartition, partitioner, f) {}
+
+  final class Holder[P, I, O](
+      var incoming: I,
+      var outgoing: Try[O],
+      val partition: P,
+      val cb: AsyncCallback[Holder[P, I, O]])
+      extends (Try[O] => Unit) {
+    private var cachedSupervisionDirective: OptionVal[Supervision.Directive] = OptionVal.None
+
+    def supervisionDirectiveFor(decider: Supervision.Decider, ex: Throwable): Supervision.Directive =
+      cachedSupervisionDirective match {
+        case OptionVal.Some(d) => d
+        case _ =>
+          val d = decider(ex)
+          cachedSupervisionDirective = OptionVal.Some(d)
+          d
+      }
+
+    def clearIncoming(): Unit = {
+      incoming = null.asInstanceOf[I]
+    }
+
+    def setOutgoing(t: Try[O]): Unit = {
+      outgoing = t
+    }
+
+    override def apply(t: Try[O]): Unit = {
+      setOutgoing(t)
+      cb.invoke(this)
+    }
+
+    override def toString: String = s"Holder($incoming, $outgoing, $partition)"
+  }
+
+  val NotYetThere = MapAsync.NotYetThere
+}
+
+private[akka] sealed abstract case class MapAsyncPartitioned[In, Out, Partition] private (
```

Comment:
`final` rather than `sealed`, not intended for internal extension. Also, `/** internal API */` and `@InternalApi` like the other operators

Reply:
Can make it final at the cost of either double-checking `parallelism` and `perPartition` or allowing instances which make no sense to be constructed.